server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -90,18 +90,29 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
         ensureNoPre019State(); //TODO remove this check, it's Elasticsearch version 7 already
         ensureAtomicMoveSupported(); //TODO move this check to NodeEnvironment, because it's related to all types of metadata
         upgradeMetaData(metaDataIndexUpgradeService, metaDataUpgrader);
-        profileLoadMetaData();
+        initializeClusterState(ClusterName.CLUSTER_NAME_SETTING.get(settings));
+        incrementalWrite = false;
     }
 
-    private void profileLoadMetaData() throws IOException {
-        if (isMasterOrDataNode()) {
-            long startNS = System.nanoTime();
-            metaStateService.loadFullState();
-            logger.debug("took {} to load state", TimeValue.timeValueMillis(TimeValue.nsecToMSec(System.nanoTime() - startNS)));
-        }
+    private void initializeClusterState(ClusterName clusterName) throws IOException {
+        long startNS = System.nanoTime();
+        Tuple<Manifest, MetaData> manifestAndMetaData = metaStateService.loadFullState();
+        previousManifest = manifestAndMetaData.v1();
+
+        previousClusterState = ClusterState.builder(clusterName)
+                .version(previousManifest.getClusterStateVersion())
+                .metaData(manifestAndMetaData.v2()).build();
+
+        logger.debug("took {} to load state", TimeValue.timeValueMillis(TimeValue.nsecToMSec(System.nanoTime() - startNS)));
     }
 
-    private void upgradeMetaData(MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader)
+    public void setLocalNode(DiscoveryNode localNode) {
```

Comment:
can you add an assertion that this is only called once?

Reply:
https://github.com/elastic/elasticsearch/pull/35819/commits/35d38872847458bd2170c4003babc5063c78c2e1